### ISSUE TITLE
Can values.yaml be used to put privately-hosted pip packages on Dask workers?

### ISSUE THREAD

**Issue by FelixVita** [dask/helm-chart#42]
I would like to deploy a dask-kubernetes helm chart that uses a **pip package from a private pypi repo**. Is this possible?

My current approach: To deploy the chart I use `helm install mydask dask/dask -f values.yaml`, and the relevant section of my yaml file looks like this:
```yaml
worker:
  env:
    - name: EXTRA_PIP_PACKAGES
      value: s3fs dask-ml myprivatepkg --extra-index-url=https://pypi.mydomain.de/token
```
I am able to deploy the helm chart, connect to the cluster, and successfully perform computations. To obtain a list of packages installed on the cluster, I use a custom function whose output is similar to that of 'pip list', but as a python function that can be submitted to the cluster:
```python
from distributed import Client
client = Client("<cluster_scheduler_ip>")
# Custom function for obtaining a list of installed packages
def list_local_packages():
    import pkg_resources
    installed_packages = pkg_resources.working_set
    installed_packages_list = sorted(["%s==%s" % (i.key, i.version) for i in installed_packages])
    return installed_packages_list
# Submit custom function to remote scheduler
client.submit(list_local_packages).result()
```
From this I obtain a list where I can see that s3fs and dask-ml were indeed installed, but **not** myprivatepkg.

If I deploy dask without the yaml, then neither s3fs nor dask-ml get installed either, so clearly the EXTRA_PIP_PACKAGES section is doing *something*.  But it seems it cannot handle the `--extra-index-url` option, even though it can handle other options like `--upgrade` just fine.

*If this feature is not implemented, is there any other approach to put privately-hosted pip packages onto the workers of a helm-deployed dask-kubernetes instance?*

**Comment by jacobtomlinson** [comment]
> I would like to deploy a dask-kubernetes helm chart

Just a quick bit of clarification before we get into discussing your issue. The [Dask Helm Chart](https://github.com/dask/helm-chart) and [dask-kubernetes](https://github.com/dask/dask-kubernetes) are two separate projects with different purposes. 

The helm chart allows you to install a jupyter server and long running Dask cluster on kubernetes. The dask-kubernetes project allows you to create ephemeral ad-hoc Dask clusters from within your Python session.

Sometimes this can be a bit confusing and folks end up on the wrong set of documentation. Just wanted to give you a heads up.

> My current approach: To deploy the chart I use helm install mydask dask/dask -f values.yaml, and the relevant section of my yaml file looks like this:

This is the correct approach to specify extra pip packages. The Dask helm chart uses the [Dask Docker image](https://github.com/dask/dask-docker) `daskdev/dask` by default. This is the container image that your environment variables are being passed to. [If you check out the `prepare.sh` script](https://github.com/dask/dask-docker/blob/master/base/prepare.sh#L24-L26) you can see where this environment variable is being handled.

As you can see we are just passing the contents of this environment variable to the `pip install` command. Therefore it is up to you to pass valid options here.

I recommend that you get a shell into one of your workers and experiment with your pip install command and continue debugging there. You can do this quickly by running `kubectl get pods` to list the pods and then run `kubectl exec -it <pod name> bash`.

**Comment by FelixVita** [comment]
Hello Jacob,

Thanks to your suggestion to use `kubectl exec -it <pod name> bash` to get a shell into one of my workers, I managed to resolve the issue. 

While in the shell of a worker, I discovered that calling `pip install` with the `--extra-index-url` option actually worked totally fine, and I was able to successfully download *myprivatepkg* from my private repo. The installation, however, failed. Turns out that what was causing the installation of my pip pkg to fail was that one of *myprivatepkg*'s dependencies required **build-essential** from the debian package repo. 

At that point I scratched my head for a while trying to figure out the best way to install extra apt packages, but then I noticed that the `prepare.sh` script (which you had linked to) had an if-statement containing the phrase `EXTRA_APT_PACKAGES`, which seemed to suggest that this was something I could use in my `values.yaml` file in a manner similar to that of `EXTRA_PIP_PACKAGES`. So I modified my yaml file like so:

```yaml
worker:
  env:
    - name: EXTRA_APT_PACKAGES
      value: build-essential
    - name: EXTRA_PIP_PACKAGES
      value: s3fs dask-ml myprivatepkg --extra-index-url=https://pypi.mydomain.de/token
```

And that fixed my issue! *Myprivatepkg* now gets installed on all workers and I can import it and use it :)

Thank you!
